File: server/app/api/v1/routes/growth_plans.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.future import select
from sqlalchemy.orm import selectinload
from typing import List

from app.api.v1.deps import get_current_user, get_session
from app.models.user import User, UserRole
from app.models.growth_week import GrowthWeek
from app.models.week_task import WeekTask
from app.schemas.growth_plan import GrowthPlanCreate, GrowthPlan as GrowthPlanSchema
from app.services.growth_plan_service import GrowthPlanService

router = APIRouter(tags=["growth-plans"])


def require_faculty_user(current_user: User = Depends(get_current_user)) -> User:
    if current_user.role != UserRole.FACULTY:
        raise HTTPException(status_code=403, detail="Only faculty users can access this endpoint")
    if not current_user.faculty_profile:
        raise HTTPException(status_code=400, detail="User has no faculty profile")
    return current_user

@router.post("/", response_model=GrowthPlanSchema)
async def create_growth_plan(
    plan_in: GrowthPlanCreate,
    current_user: User = Depends(require_faculty_user),
    db: AsyncSession = Depends(get_session)
):
    service = GrowthPlanService(db)
    try:
        plan = await service.create_plan(current_user.faculty_profile.id, plan_in)
        return plan
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except RuntimeError as e:
        raise HTTPException(status_code=status.HTTP_503_SERVICE_UNAVAILABLE, detail=str(e))

@router.get("/me", response_model=GrowthPlanSchema)
async def get_my_active_plan(
    current_user: User = Depends(require_faculty_user),
    db: AsyncSession = Depends(get_session)
):
    service = GrowthPlanService(db)
    plan = await service.get_active_plan(current_user.faculty_profile.id)
    if not plan:
        raise HTTPException(status_code=404, detail="No active growth plan found")
    return plan

@router.delete("/me")
async def reset_my_plan(
    current_user: User = Depends(require_faculty_user),
    db: AsyncSession = Depends(get_session)
):
    service = GrowthPlanService(db)
    await service.reset_plan(current_user.faculty_profile.id)
    return {"status": "success"}

@router.post("/weeks/{week_id}/complete")
async def complete_week(
    week_id: str,
    current_user: User = Depends(require_faculty_user),
    db: AsyncSession = Depends(get_session)
):
    week_result = await db.execute(
        select(GrowthWeek)
        .where(GrowthWeek.id == week_id)
        .options(selectinload(GrowthWeek.plan))
    )
    week = week_result.scalar_one_or_none()
    if not week:
        raise HTTPException(status_code=404, detail="Week not found or already completed")
    if week.plan.faculty_id != current_user.faculty_profile.id:
        raise HTTPException(status_code=403, detail="Cannot modify another faculty's growth plan")

    service = GrowthPlanService(db)
    success = await service.complete_week(week_id)
    if not success:
        raise HTTPException(status_code=404, detail="Week not found or already completed")
    return {"status": "success"}

@router.patch("/tasks/{task_id}")
async def update_task_status(
    task_id: str,
    done: bool,
    current_user: User = Depends(require_faculty_user),
    db: AsyncSession = Depends(get_session)
):
    task_result = await db.execute(
        select(WeekTask)
        .where(WeekTask.id == task_id)
        .options(selectinload(WeekTask.week).selectinload(GrowthWeek.plan))
    )
    task = task_result.scalar_one_or_none()
    if not task:
        raise HTTPException(status_code=404, detail="Task not found")
    if task.week.plan.faculty_id != current_user.faculty_profile.id:
        raise HTTPException(status_code=403, detail="Cannot modify another faculty's growth plan")

    service = GrowthPlanService(db)
    success = await service.update_task_status(task_id, done)
    if not success:
        raise HTTPException(status_code=404, detail="Task not found")
    return {"status": "success"}

@router.get("/", response_model=List[GrowthPlanSchema])
async def list_growth_plans(
    skip: int = 0,
    limit: int = 100,
    current_user: User = Depends(get_current_user), # Admin or Faculty?
    db: AsyncSession = Depends(get_session)
):
    # For simplicity, admin can see all
    if current_user.role != UserRole.ADMIN:
         raise HTTPException(status_code=403, detail="Only admins can list all growth plans")
         
    service = GrowthPlanService(db)
    plans = await service.list_plans(skip=skip, limit=limit)
    return plans

File: client/src/lib/api/growthPlans.ts
import http from './http'
import { SkillDomain } from './skills'

export enum GrowthPlanStatus {
    ACTIVE = "ACTIVE",
    COMPLETED = "COMPLETED",
    RESET = "RESET",
}

export interface WeekTask {
    id: string;
    week_id: string;
    label: string;
    done: boolean;
}

export interface GrowthWeek {
    id: string;
    plan_id: string;
    week_number: number;
    title: string;
    required_practice_count: number;
    required_min_avg_score: number;
    completed_practice_count: number;
    avg_score_for_week: number;
    completed: boolean;
    completed_at?: string;
    tasks: WeekTask[];
}

export interface GrowthPlan {
    id: string;
    faculty_id: string;
    domain: SkillDomain;
    target_skill: string;
    current_level: number;
    target_level: number;
    weekly_hours: number;
    status: GrowthPlanStatus;
    progress_percentage: number;
    created_at: string;
    reset_at?: string;
    weeks: GrowthWeek[];
}

export interface GrowthPlanCreate {
    domain: SkillDomain;
    target_skill: string;
    current_level: number;
    target_level: number;
    weekly_hours: number;
}

export const growthPlansApi = {
    createPlan: async (planData: GrowthPlanCreate): Promise<GrowthPlan> => {
        const response = await http.post<GrowthPlan>('/api/v1/growth-plans/', planData);
        return response.data;
    },

    getMyActivePlan: async (): Promise<GrowthPlan> => {
        const response = await http.get<GrowthPlan>('/api/v1/growth-plans/me');
        return response.data;
    },

    updateTaskStatus: async (taskId: string, done: boolean): Promise<{ status: string }> => {
        const response = await http.patch<{ status: string }>(`/api/v1/growth-plans/tasks/${taskId}`, null, {
            params: { done }
        });
        return response.data;
    },

    listPlans: async (skip: number = 0, limit: number = 100): Promise<GrowthPlan[]> => {
        const response = await http.get<GrowthPlan[]>('/api/v1/growth-plans/', {
            params: { skip, limit }
        });
        return response.data;
    },

    hardReset: async (): Promise<{ status: string }> => {
        const response = await http.delete<{ status: string }>('/api/v1/growth-plans/me');
        return response.data;
    },

    completeWeek: async (weekId: string): Promise<{ status: string }> => {
        const response = await http.post<{ status: string }>(`/api/v1/growth-plans/weeks/${weekId}/complete`);
        return response.data;
    }
}
